src/device-config.ts

```diff
@@ -0,0 +1,512 @@
+		let reboot = false;
+
+		// If the legacy lock override is used, place it as the new variable
+		if (checkTruthy(target['OVERRIDE_LOCK'])) {
+			target['SUPERVISOR_OVERRIDE_LOCK'] = target['OVERRIDE_LOCK'];
+		}
+
+		_.each(
+			DeviceConfig.configKeys,
+			({ envVarName, varType, rebootRequired }, key) => {
+				// Test if the key is different
+				if (
+					!DeviceConfig.configTest(
+						varType,
+						current[envVarName],
+						target[envVarName],
+					)
+				) {
+					// Save the change if it is
+					configChanges[key] = target[envVarName];
+					humanReadableConfigChanges[envVarName] = target[envVarName];
+					reboot = reboot || (rebootRequired != null ? rebootRequired : false);
+				}
+			},
+		);
+
+		if (!_.isEmpty(configChanges)) {
+			steps.push({
+				action: 'changeConfig',
+				target: configChanges,
+				humanReadableTarget: humanReadableConfigChanges,
+				rebootRequired: reboot,
+			});
+		}
+
+		// Check for special case actions for the VPN
+		if (
+			!checkTruthy(unmanaged || false) &&
+			!_.isEmpty(target['SUPERVISOR_VPN_CONTROL']) &&
+			DeviceConfig.checkBoolChanged(current, target, 'SUPERVISOR_VPN_CONTROL')
+		) {
+			steps.push({
+				action: 'setVPNEnabled',
+				target: target['SUPERVISOR_VPN_CONTROL'],
+			});
+		}
+
+		// Do we need to change the boot config?
+		if (this.bootConfigChangeRequired(backend, current, target)) {
+			steps.push({
+				action: 'setBootConfig',
+				target,
+			});
+		}
+
+		if (_.isEmpty(steps) && this.rebootRequired) {
+			steps.push({
+				action: 'reboot',
+			});
+		}
+
+		return steps;
+	}
+
+	public executeStepAction(step: ConfigStep, opts: DeviceActionExecutorOpts) {
+		if (step.action !== 'reboot') {
+			return this.actionExecutors[step.action](step, opts);
+		}
+	}
+
+	public isValidAction(action: string): boolean {
+		return _.includes(_.keys(this.actionExecutors), action);
+	}
+
+	private async getBootConfig(
+		backend: DeviceConfigBackend | null,
+	): Promise<EnvVarObject> {
+		if (backend == null) {
+			return {};
+		}
+		const conf = await backend.getBootConfig();
+		return configUtils.bootConfigToEnv(backend, conf);
+	}
+
+	private async setBootConfig(
+		backend: DeviceConfigBackend | null,
+		target: Dictionary<string>,
+	) {
+		if (backend == null) {
+			return false;
+		}
+
+		const conf = configUtils.envToBootConfig(backend, target);
+		this.logger.logSystemMessage(
+			`Applying boot config: ${JSON.stringify(conf)}`,
+			{},
+			'Apply boot config in progress',
+		);
+
+		try {
+			await backend.setBootConfig(conf);
+			this.logger.logSystemMessage(
+				`Applied boot config: ${JSON.stringify(conf)}`,
+				{},
+				'Apply boot config success',
+			);
+			this.rebootRequired = true;
+			return true;
+		} catch (err) {
+			this.logger.logSystemMessage(
+				`Error setting boot config: ${err}`,
+				{ error: err },
+				'Apply boot config error',
+			);
+			throw err;
+		}
+	}
+
+	private async getVPNEnabled(): Promise<boolean | null> {
```

Comment:
I don't see where this can return null?

Reply:
Ah it used to in the original implementation, but it didn't make any sense and I forgot to take it out of the sig - good catch!